File: aoc2025/javascript/day8.js
import { readFile } from './utilities.js';

class JunctionBox {
  constructor(x, y, z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  toString() {
    return `${this.x},${this.y},${this.z}`;
  }
}

class JunctionBoxGroup {
  constructor(connections = []) {
    this.connections = connections;
  }
}

class ShortestConnection {
  constructor(distance, from, to) {
    this.distance = distance;
    this.from = from;
    this.to = to;
  }
}

function buildJunctionBoxes(filename) {
  const lines = readFile(filename);
  const junctionBoxes = [];
  const junctionBoxConnectionIndex = new Map();

  for (const row of lines) {
    const coordinates = row.split(',');
    const x = parseInt(coordinates[0]);
    const y = parseInt(coordinates[1]);
    const z = parseInt(coordinates[2]);

    const newJunctionBox = new JunctionBox(x, y, z);
    junctionBoxes.push(newJunctionBox);
    junctionBoxConnectionIndex.set(newJunctionBox.toString(), -1);
  }

  return [junctionBoxes, junctionBoxConnectionIndex];
}

function getSortedConnections(junctionBoxes) {
  const shortestConnections = [];

  for (let i = 0; i < junctionBoxes.length; i++) {
    for (let k = i + 1; k < junctionBoxes.length; k++) {
      const junctionBox = junctionBoxes[i];
      const otherJunctionBox = junctionBoxes[k];

      const dx = junctionBox.x - otherJunctionBox.x;
      const dy = junctionBox.y - otherJunctionBox.y;
      const dz = junctionBox.z - otherJunctionBox.z;

      const distance = Math.floor(Math.sqrt(dx * dx + dy * dy + dz * dz));

      shortestConnections.push(
        new ShortestConnection(distance, junctionBox, otherJunctionBox)
      );
    }
  }

  shortestConnections.sort((a, b) => a.distance - b.distance);
  return shortestConnections;
}

function buildJunctionBoxConnectionsPart1(
  shortestConnections,
  junctionBoxConnectionIndex,
  junctionBoxes,
  limit
) {
  const connections = [];
  let connectionsMade = 0;

  for (const connection of shortestConnections) {
    if (connectionsMade >= limit) {
      break;
    }

    const fromIndex = junctionBoxConnectionIndex.get(connection.from.toString());
    const toIndex = junctionBoxConnectionIndex.get(connection.to.toString());

    connectionsMade++;
    if (fromIndex !== -1 && fromIndex === toIndex) {
      continue;
    }

    if (fromIndex === -1 && toIndex === -1) {
      connections.push(new JunctionBoxGroup([connection.from, connection.to]));
      junctionBoxConnectionIndex.set(connection.from.toString(), connections.length - 1);
      junctionBoxConnectionIndex.set(connection.to.toString(), connections.length - 1);
      continue;
    }

    if (fromIndex === -1) {
      connections[toIndex].connections.push(connection.from);
      junctionBoxConnectionIndex.set(connection.from.toString(), toIndex);
      continue;
    }

    if (toIndex === -1) {
      connections[fromIndex].connections.push(connection.to);
      junctionBoxConnectionIndex.set(connection.to.toString(), fromIndex);
      continue;
    }

    let smallerIdx = fromIndex;
    let largerIdx = toIndex;
    if (connections[fromIndex].connections.length > connections[toIndex].connections.length) {
      smallerIdx = toIndex;
      largerIdx = fromIndex;
    }

    for (const box of connections[smallerIdx].connections) {
      junctionBoxConnectionIndex.set(box.toString(), largerIdx);
    }

    connections[largerIdx].connections.push(...connections[smallerIdx].connections);
  }

  for (const junctionBox of junctionBoxes) {
    if (junctionBoxConnectionIndex.get(junctionBox.toString()) === -1) {
      connections.push(new JunctionBoxGroup([junctionBox]));
      junctionBoxConnectionIndex.set(junctionBox.toString(), connections.length - 1);
    }
  }

  connections.sort((a, b) => b.connections.length - a.connections.length);

  let result = 0;
  for (let i = 0; i < Math.min(3, connections.length); i++) {
    if (result === 0) {
      result = connections[i].connections.length;
    } else {
      result *= connections[i].connections.length;
    }
  }

  return result;
}

function buildJunctionBoxConnectionsPart2(
  shortestConnections,
  junctionBoxConnectionIndex,
  junctionBoxes
) {
  const connections = [];
  let connectionsMade = 0;

  for (let i = 0; i < shortestConnections.length; i++) {
    const connection = shortestConnections[i];

    if (connections.length === 1 && connections[0].connections.length === junctionBoxes.length) {
      return shortestConnections[i - 1].from.x * shortestConnections[i - 1].to.x;
    }

    const fromIndex = junctionBoxConnectionIndex.get(connection.from.toString());
    const toIndex = junctionBoxConnectionIndex.get(connection.to.toString());

    connectionsMade++;
    if (fromIndex !== -1 && fromIndex === toIndex) {
      continue;
    }

    if (fromIndex === -1 && toIndex === -1) {
      connections.push(new JunctionBoxGroup([connection.from, connection.to]));
      junctionBoxConnectionIndex.set(connection.from.toString(), connections.length - 1);
      junctionBoxConnectionIndex.set(connection.to.toString(), connections.length - 1);
      continue;
    }

    if (fromIndex === -1) {
      connections[toIndex].connections.push(connection.from);
      junctionBoxConnectionIndex.set(connection.from.toString(), toIndex);
      continue;
    }

    if (toIndex === -1) {
      connections[fromIndex].connections.push(connection.to);
      junctionBoxConnectionIndex.set(connection.to.toString(), fromIndex);
      continue;
    }

    let smallerIdx = fromIndex;
    let largerIdx = toIndex;
    if (connections[fromIndex].connections.length > connections[toIndex].connections.length) {
      smallerIdx = toIndex;
      largerIdx = fromIndex;
    }

    for (const box of connections[smallerIdx].connections) {
      junctionBoxConnectionIndex.set(box.toString(), largerIdx);
    }

    connections[largerIdx].connections.push(...connections[smallerIdx].connections);

    connections.splice(smallerIdx, 1);

    for (const [key, idx] of junctionBoxConnectionIndex.entries()) {
      if (idx > smallerIdx) {
        junctionBoxConnectionIndex.set(key, idx - 1);
      }
    }
  }

  return 0;
}

function day8() {
  let [junctionBoxes, junctionBoxConnectionIndex] = buildJunctionBoxes('day8.part1');
  const connections = getSortedConnections(junctionBoxes);

  let start = performance.now();
  const part1Results = buildJunctionBoxConnectionsPart1(
    connections,
    junctionBoxConnectionIndex,
    junctionBoxes,
    1000
  );
  const part1Time = ((performance.now() - start) / 1000).toFixed(4);

  [junctionBoxes, junctionBoxConnectionIndex] = buildJunctionBoxes('day8.part1');
  start = performance.now();
  const part2Results = buildJunctionBoxConnectionsPart2(
    connections,
    junctionBoxConnectionIndex,
    junctionBoxes
  );
  const part2Time = ((performance.now() - start) / 1000).toFixed(4);

  console.log(`\nDay 8:`);
  console.log(`  Part 1: ${part1Results} (${part1Time}s)`);
  console.log(`  Part 2: ${part2Results} (${part2Time}s)`);
}

day8();

File: aoc2025/python/day8.py
import time
import math
from utilities import read_file


class JunctionBox:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y and self.z == other.z

    def __hash__(self):
        return hash((self.x, self.y, self.z))

    def __repr__(self):
        return f"JunctionBox({self.x}, {self.y}, {self.z})"


class JunctionBoxGroup:
    def __init__(self, connections=None):
        self.connections = connections if connections else []


class ShortestConnection:
    def __init__(self, distance, from_box, to_box):
        self.distance = distance
        self.from_box = from_box
        self.to_box = to_box


def build_junction_boxes(filename):
    lines = read_file(filename)
    junction_boxes = []
    junction_box_connection_index = {}

    for row in lines:
        coordinates = row.split(",")
        x = int(coordinates[0])
        y = int(coordinates[1])
        z = int(coordinates[2])

        new_junction_box = JunctionBox(x, y, z)
        junction_boxes.append(new_junction_box)
        junction_box_connection_index[new_junction_box] = -1

    return junction_boxes, junction_box_connection_index


def get_sorted_connections(junction_boxes):
    shortest_connections = []

    for i in range(len(junction_boxes)):
        for k in range(i + 1, len(junction_boxes)):
            junction_box = junction_boxes[i]
            other_junction_box = junction_boxes[k]

            dx = junction_box.x - other_junction_box.x
            dy = junction_box.y - other_junction_box.y
            dz = junction_box.z - other_junction_box.z

            distance = int(math.sqrt(dx * dx + dy * dy + dz * dz))

            shortest_connections.append(
                ShortestConnection(distance, junction_box, other_junction_box)
            )

    shortest_connections.sort(key=lambda c: c.distance)
    return shortest_connections


def build_junction_box_connections_part1(
    shortest_connections, junction_box_connection_index, junction_boxes, limit
):
    connections = []
    connections_made = 0

    for connection in shortest_connections:
        if connections_made >= limit:
            break

        from_index = junction_box_connection_index[connection.from_box]
        to_index = junction_box_connection_index[connection.to_box]

        connections_made += 1
        if from_index != -1 and from_index == to_index:
            continue

        if from_index == -1 and to_index == -1:
            connections.append(
                JunctionBoxGroup([connection.from_box, connection.to_box])
            )
            junction_box_connection_index[connection.from_box] = len(connections) - 1
            junction_box_connection_index[connection.to_box] = len(connections) - 1
            continue

        if from_index == -1:
            connections[to_index].connections.append(connection.from_box)
            junction_box_connection_index[connection.from_box] = to_index
            continue

        if to_index == -1:
            connections[from_index].connections.append(connection.to_box)
            junction_box_connection_index[connection.to_box] = from_index
            continue

        smaller_idx = from_index
        larger_idx = to_index
        if len(connections[from_index].connections) > len(
            connections[to_index].connections
        ):
            smaller_idx = to_index
            larger_idx = from_index

        for box in connections[smaller_idx].connections:
            junction_box_connection_index[box] = larger_idx

        connections[larger_idx].connections.extend(
            connections[smaller_idx].connections
        )

    for junction_box in junction_boxes:
        if junction_box_connection_index[junction_box] == -1:
            connections.append(JunctionBoxGroup([junction_box]))
            junction_box_connection_index[junction_box] = len(connections) - 1

    connections.sort(key=lambda g: len(g.connections), reverse=True)

    result = 0
    for i, connection in enumerate(connections):
        if i >= 3:
            break
        if result == 0:
            result = len(connection.connections)
        else:
            result *= len(connection.connections)

    return result


def build_junction_box_connections_part2(
    shortest_connections, junction_box_connection_index, junction_boxes
):
    connections = []
    connections_made = 0

    for i, connection in enumerate(shortest_connections):
        if (
            len(connections) == 1
            and len(connections[0].connections) == len(junction_boxes)
        ):
            return (
                shortest_connections[i - 1].from_box.x
                * shortest_connections[i - 1].to_box.x
            )

        from_index = junction_box_connection_index[connection.from_box]
        to_index = junction_box_connection_index[connection.to_box]

        connections_made += 1
        if from_index != -1 and from_index == to_index:
            continue

        if from_index == -1 and to_index == -1:
            connections.append(
                JunctionBoxGroup([connection.from_box, connection.to_box])
            )
            junction_box_connection_index[connection.from_box] = len(connections) - 1
            junction_box_connection_index[connection.to_box] = len(connections) - 1
            continue

        if from_index == -1:
            connections[to_index].connections.append(connection.from_box)
            junction_box_connection_index[connection.from_box] = to_index
            continue

        if to_index == -1:
            connections[from_index].connections.append(connection.to_box)
            junction_box_connection_index[connection.to_box] = from_index
            continue

        smaller_idx = from_index
        larger_idx = to_index
        if len(connections[from_index].connections) > len(
            connections[to_index].connections
        ):
            smaller_idx = to_index
            larger_idx = from_index

        for box in connections[smaller_idx].connections:
            junction_box_connection_index[box] = larger_idx

        connections[larger_idx].connections.extend(
            connections[smaller_idx].connections
        )

        connections.pop(smaller_idx)

        for box in list(junction_box_connection_index.keys()):
            if junction_box_connection_index[box] > smaller_idx:
                junction_box_connection_index[box] -= 1

    return 0


def day8():
    junction_boxes, junction_box_connection_index = build_junction_boxes("day8.part1")
    connections = get_sorted_connections(junction_boxes)

    start = time.time()
    part1_results = build_junction_box_connections_part1(
        connections, junction_box_connection_index, junction_boxes, 1000
    )
    part1_time = time.time() - start

    junction_boxes, junction_box_connection_index = build_junction_boxes("day8.part1")
    start = time.time()
    part2_results = build_junction_box_connections_part2(
        connections, junction_box_connection_index, junction_boxes
    )
    part2_time = time.time() - start

    print(f"Day 8:")
    print(f"  Part 1: {part1_results} ({part1_time:.4f}s)")
    print(f"  Part 2: {part2_results} ({part2_time:.4f}s)")


if __name__ == "__main__":
    day8()
